substantial/types.py

```diff
@@ -75,23 +79,52 @@ def linear(self, retries_left: int) -> timedelta:
         )
 
 
+@dataclass
+class CompensationStrategy:
+    compensate_fn: Optional[Callable[[], Any]] = None
+    max_compensation_attemps: int = 3
+
+
+class CompensationFailed(BaseException):
+    def __init__(
+        self,
+        original_error: Exception,
+        compensation_error: Optional[Exception] = None,
+    ):
+        self.original_error = original_error
+        self.compensation_error = compensation_error
+
+
 @dataclass
 class ValueEval:
     lambda_fn: Callable[[], Any]
-    timeout: Union[int, None]
+    timeout: Union[float, None]
     retry_strategy: Union[RetryStrategy, None]
+    compensate_fn: Optional[Callable[[], Any]]
+    max_compensation_attempts: int = 3
+
+    async def exec(
+        self,
+        ctx,
+        save_id,
+        counter,
+        compensation_stack: Optional[List[Callable[[], Any]]] = None,
```

Comment:
Why have a new stack at every save exec?

Reply:
cause of this comment https://github.com/zifeo/substantial/pull/10#discussion_r1875793063

Reply:
Ah, I think you have misunderstood it, my bad for not clarifying it better.

That was because I assumed you'd use a owned reference of the stack, and I just pointed out that it was shared across all callers. I already mentioned a scheme to counter that (maybe a dict whose key differentiate callers). But I still insisted to not do that as it was not very portable (remember this will be ported into metatype).

Now you have a stack that is renewed for every save and whose reference is only kept within the exec function. Meaning your compensation stack will always be at most 1 it it will **only** contain the compensate for that operation.